refactor!: introduce new update algorithm based on `git merge`

sisp opened the pull request:
I have introduced a new update algorithm that is centered around the high-level `git merge` command, performing a 3-way merge of a fresh project based on the new template version into the current project with the common ancestor being a fresh project based on the old template version. This is – in a nutshell – what the previous update algorithm was doing using a sequence of low-level Git commands and other hacks. With this refactoring, I'm hoping to simplify the update algorithm implementation, reduce the risk of missed edge cases by relying on a battle-tested high-level command, and paving the way for new opportunities like improving merge conflict handling through custom merge drivers like <https://mergiraf.org> and supporting an update algorithm variant that does not rely on replaying a fresh project based on the old template version (see #1170).

I have marked this PR as a draft because this is a significant (and breaking, see below for more details) change that needs to be thoroughly reviewed and shouldn't be rushed.

According to my research, Copier's update algorithm has a long history, starting with #106 and evolving from a 2-way merge to a 3-way merge with merge conflicts stored first in `.rej` files and later also available as inline markers. Many edge cases have shown up along the way and their fixes contributed to the current complexity. At some point, I realized that the current algorithm is essentially a 3-way merge, implemented with a sequence of low-level commands and hacks. But `git merge` also performs a 3-way merge with just a single command, given that the Git graph contains three commits with the appropriate relationship. This Git graph does not exist naturally in our case, as a Copier template needs to be instantiated before updating a project, and these template instances are not tracked via Git. But it is possible to construct this relationship in a synthetic Git graph which consists of the three commits necessary to perform the 3-way merge. `git merge` can be applied on this synthetic graph, and with some tricks it looks like a familiar merge. Thanks to using plain `git merge`, [delicate Git index manipulation](https://github.com/copier-org/copier/blob/de88305564e99a698289ba2e108edf410f85bb78/copier/_main.py#L1377-L1406) to induce a mid-merge state for merge conflicts, especially to enable [visual merge conflicts in IDEs like VS Code](https://github.com/copier-org/copier/issues/1350), becomes obsolete, aborting a Copier update becomes a simple `git merge --abort` command, and edge cases like [add/add merge conflicts](https://github.com/copier-org/copier/pull/1587) are handled without special treatment. In a way, `copier copy` is like `git clone`, and `copier update` is like `git merge` – but extended to handling parametrized starters with some extra sugar.

This PR introduces a few breaking changes:

- `git merge` can only produce inline conflict markers but no `.rej` files, so I've removed the `conflict` setting. I suppose that nobody uses `.rej` files anymore, as inline conflict markers have been the default setting for a while and are familiar to Git users.
- `git merge` does not support a configurable context size, so I've removed the `context_lines` setting.
- `git merge` does not support full customization of the conflict maker labels. To compensate for this limitation at least a little bit, a temporary ref `copier/after-updating` points to the commit of the fresh project based on the new template version, which is used as the conflict marker label for "incoming" changes. The conflict marker label for the current state is `HEAD`.
- `.gitattributes` settings affect Copier's internal `git merge` call. This may have (unexpected) side effects which are difficult to anticipate. But it may be exactly what we want, glancing at a custom merge driver like <https://mergiraf.org> again.

If we decide to (eventually) merge this PR, I suggest to slip in deprecation warnings first, release them, and wait for a little while before creating a new major release to give Copier users a smoother transition experience.

WDYT, @copier-org/maintainers?

Additionally looping in some contributors who have been active around the update algorithm: @tpluscode @tguillemot @barrywhart @thurse93

sisp committed:
feat(updating)!: introduce new update algorithm based on `git merge`

BREAKING CHANGE: The `-o,--conflict` flag is removed, as `git merge`
only supports inline conflicts. Omit the flag and corresponding API
parameter as of now.
BREAKING CHANGE: The `-c,--context-lines` flag is removed, as
`git merge` does not support configurable context sizes. Omit the flag
and corresponding API parameter as of now.
BREAKING CHANGE: Inline conflict marker labels are different, as
`git merge` does not support their customization.
BREAKING CHANGE: `gitattributes` settings affect Copier's internal
`git merge` call.

yajo reviewed (approved):
I'm pretty amazed with the new algorithm. I have not tested it yet, but with reduced low-level plumbing and half of the lines, all tests pass! At least, except the ones you're inherently breaking.

I suggest to add `BREAKING CHANGE:` trailers to the git commit message listing specific things that break, so they appear in the changelog once released.

Thank you! ❤️

sisp commented:
We might need to run pre-update migrations also on the fresh copy of the old template version – see https://github.com/copier-org/copier/discussions/2612#discussioncomment-16545129. But this needs some deep thinking to make sure it's _generally_ correct.

sisp committed:
fix(updating)!: do not run tasks after project update

Running tasks after an update mixes concerns with post-copy tasks and
post-update migration tasks.

BREAKING CHANGE: Copier no longer runs tasks after updating a project.
Use post-update migration tasks instead.

sisp committed:
feat(updating)!: apply pre-update migrations on fresh project from current template version

BREAKING CHANGE: Pre-update migrations are now run on the fresh project
generated from the current template version. In most cases, this should
only improve update quality by reducing merge conflicts, but it is
possible that this change will break template updates in rare cases.
Thus, this change is marked as breaking out of an abundance of caution.
